{"cells":[{"cell_type":"markdown","id":"789d6b06","metadata":{"id":"789d6b06"},"source":["# SCRIPT FOR ANALYSIS OF CELLS ANALYZED WITH Xenium"]},{"cell_type":"markdown","id":"374ccf62","metadata":{"id":"374ccf62"},"source":["This notebook is used to identify cell populations from a Xenium dataset. The analysis is based on Scanpy (documentation in  https://scanpy.readthedocs.io/en/stable/index.html) and Squidpy (documentation in https://squidpy.readthedocs.io/en/stable/)"]},{"cell_type":"markdown","id":"74feca71","metadata":{"id":"74feca71"},"source":["## Loading the packages"]},{"cell_type":"markdown","id":"335c72e4","metadata":{"id":"335c72e4"},"source":["We import the packages needed for it. If something fails, means that we haven't installed something in the environment"]},{"cell_type":"code","execution_count":null,"id":"508e75f8","metadata":{"id":"508e75f8"},"outputs":[],"source":["import numpy as np\n","import pandas as pd\n","import scanpy as sc\n","import seaborn as sns\n","import matplotlib.pyplot as plt\n","#from xb.plotting import map_of_clusters\n","#import squidpy as sq"]},{"cell_type":"code","execution_count":null,"id":"2a03ea5c","metadata":{"id":"2a03ea5c"},"outputs":[],"source":["def map_of_clusters(adata,key='leiden',clusters='all',size=8,background='white',figuresize=(10,7),save=None,format='pdf'):\n","\n","      #PARAMETERS TO MODIFY:\n","     #-key: the terms in adata.obs that you want to plot\n","     #-clusters:'all' for plotting all clusters in a single plot, 'individual': for plots of individual genes, or ['3','5'] (your groups\n","     #          between square brackets to plot only some clusters\n","     #-size: to change the size of your spots\n","     #-background: to change the color of the background\n","     #-figuresize: to specify the size of your figure\n","     #-save: if you want to save your figure, give the PATH of the folder where you want to save it\n","     #-format: specify the format in which you want to save your figure\n","     try:\n","         adata.obs[key]=adata.obs[key].astype(int)\n","         colors=dict(zip(np.unique(adata.obs[key]),adata.uns[key+'_colors']))\n","     except:\n","         colors=dict(zip(np.unique(adata.obs[key]),adata.uns[key+'_colors']))\n","     #cl.apply(lambda x: colors[x])\n","     plt.rcParams['figure.facecolor'] = background\n","     if clusters=='all':\n","         cl=adata.obs[key]\n","         plt.figure(figsize=figuresize)\n","         figa=plt.scatter(x=adata.obs.X,y=adata.obs.Y,c=cl.apply(lambda x: colors[x]),s=size,linewidths=0, edgecolors=None)\n","         plt.axis('off')\n","         if not save==None:\n","             plt.savefig(save +'/map_all_clusters_'+str(size)+'_'+background+'_'+key+'.'+format)\n","     elif clusters=='individual':\n","         cl=adata.obs[key]\n","         for each in adata.obs[key].unique():\n","             adatasub=adata[adata.obs[key]==each]\n","             plt.figure(figsize=figuresize)\n","             plt.scatter(x=adata.obs.X,y=adata.obs.Y,c='grey',s=size/5,linewidths=0, edgecolors=None)\n","             cl=adatasub.obs[key]\n","             plt.scatter(x=adatasub.obs.X,y=adatasub.obs.Y,c=cl.apply(lambda x: colors[x]),s=size,linewidths=0, edgecolors=None)\n","             plt.axis('off')\n","             plt.title('Group: '+ str(each))\n","             if not save==None:\n","                 plt.savefig(save +'/map_inidivdual_cluster_'+str(each)+'_'+str(size)+background+'_'+key+'.'+format)\n","     else:\n","         adatasub=adata[adata.obs[key].isin(clusters)]\n","         plt.figure(figsize=figuresize)\n","         plt.scatter(x=adata.obs.X,y=adata.obs.Y,c='grey',s=size/5,linewidths=0, edgecolors=None)\n","         cl=adatasub.obs[key]\n","         plt.scatter(x=adatasub.obs.X,y=adatasub.obs.Y,c=cl.apply(lambda x: colors[x]),s=size,linewidths=0, edgecolors=None)\n","         plt.axis('off')\n","         plt.legend()\n","         if not save==None:\n","                 s=''\n","                 for element in clusters:\n","                     s=s+str(element)\n","                 print(s)\n","                 plt.savefig(save +'/map_group_of_clusters_'+str(s)+'_'+str(size)+background+'_'+key+'.'+format)\n"," #        plt.title('Group: '+ paste(clusters))"]},{"cell_type":"markdown","id":"efcbf1ef-ce9c-44db-bf34-e3b37104ab12","metadata":{"id":"efcbf1ef-ce9c-44db-bf34-e3b37104ab12"},"source":["Importing the data and creating an anndata object"]},{"cell_type":"markdown","id":"c282aae4","metadata":{"id":"c282aae4"},"source":["We import the adata object for each of the three technical replicates. We remove articitial/control genes and we add a tag to every cell to distinguish the replicate they are comming from"]},{"cell_type":"code","execution_count":null,"id":"80bd3245","metadata":{"id":"80bd3245"},"outputs":[],"source":["outpath=r'/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/' # path where to save the nuclei adata\n","adata=sc.read(outpath+'/adata_input.h5ad')"]},{"cell_type":"markdown","id":"d117d758","metadata":{"id":"d117d758"},"source":["We now save the raw data in adata.raw and apply normalization and log-transformation to minimize the effect of outliers. We also scale the data to give the same importance to all genes in clustering"]},{"cell_type":"code","execution_count":null,"id":"5ded2ea1","metadata":{"id":"5ded2ea1"},"outputs":[],"source":["adata=adata[:,[g for g in adata.var.index if 'UnassignedCodeword' not in g ]]\n","adata=adata[:,[g for g in adata.var.index if 'NegControlCodeword' not in g ]]"]},{"cell_type":"code","execution_count":null,"id":"14682e8e","metadata":{"id":"14682e8e"},"outputs":[],"source":["adata.raw=adata\n","sc.pp.filter_cells(adata,min_genes=8)\n","sc.pp.filter_cells(adata,min_counts=20)\n","adata.layers['raw']=adata.X\n","sc.pp.normalize_total(adata, target_sum=None)\n","sc.pp.log1p(adata)\n","#sc.pp.scale(adata) ####note that we are scaling"]},{"cell_type":"markdown","id":"feb77aac","metadata":{"id":"feb77aac"},"source":["# Dimensional reduction and clustering"]},{"cell_type":"markdown","id":"a5456887","metadata":{"id":"a5456887"},"source":["We apply principal component analysis to find the appropiate nubmer of components to use to perform dimensionality reduction and clustering"]},{"cell_type":"code","execution_count":null,"id":"87f88bbe","metadata":{"id":"87f88bbe"},"outputs":[],"source":["adata"]},{"cell_type":"code","execution_count":null,"id":"41caef05","metadata":{"id":"41caef05"},"outputs":[],"source":["sc.tl.pca(adata, svd_solver='arpack')\n","plt.rcParams['figure.facecolor'] = 'white'\n","sc.pl.pca_variance_ratio(adata, log=False,n_pcs=50)"]},{"cell_type":"markdown","id":"fa57958b","metadata":{"id":"fa57958b"},"source":["We select to go with the top X first principal components to look for the nearest neighbors. We are going to represent our data using graph-based methods, like UMAP. For this, we first need to create a graph by linking each cell to the n closest neighbors (n_neighbors) in the dataset. We can select the numbers of principal components to use"]},{"cell_type":"code","execution_count":null,"id":"c5aed22d","metadata":{"id":"c5aed22d"},"outputs":[],"source":["sc.pp.neighbors(adata, n_neighbors=12, n_pcs=0)"]},{"cell_type":"markdown","id":"51bc254c","metadata":{"id":"51bc254c"},"source":["We can, then, run clustering, by calling either leiden or louvain clustering. By changing the parameter \"resolution\" we will generate more or less clusters. Higher resolution means more clusters. The clusters (groups of cells) are indepent from its visualization, which we will create using UMAP later on"]},{"cell_type":"code","execution_count":null,"id":"37f9a20d","metadata":{"id":"37f9a20d"},"outputs":[],"source":["sc.tl.umap(adata,min_dist=0.1)"]},{"cell_type":"code","execution_count":null,"id":"e4774a4e","metadata":{"id":"e4774a4e"},"outputs":[],"source":["#sc.tl.leiden(adata,resolution=2.0,key_added='leiden_2_0')\n","sc.tl.leiden(adata,resolution=1.0,key_added='leiden_1_0')\n","#sc.tl.leiden(adata,resolution=0.6,key_added='leiden_0_6')"]},{"cell_type":"code","execution_count":null,"id":"e5618ed7","metadata":{"id":"e5618ed7"},"outputs":[],"source":["sc.tl.leiden(adata,resolution=1.0,key_added='leiden_1_0')"]},{"cell_type":"markdown","id":"56d22752","metadata":{"id":"56d22752"},"source":["To VISUALIZE the diversity of our cells, We perfrom UMAP. The closest two spots are in UMAP, the more similar they will be. With the comand sc.tl.umap. we CALCULATE the UMAP embedding. With sc.pl.umap will VISUALIZE the UMAP."]},{"cell_type":"code","execution_count":null,"id":"90264401","metadata":{"id":"90264401"},"outputs":[],"source":["adata.write(outpath+'/adata_after_leiden_noscale.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"2d0dc584","metadata":{"id":"2d0dc584"},"outputs":[],"source":["adata=sc.read(outpath+'/adata_after_leiden_noscale.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"40da97d1","metadata":{"id":"40da97d1"},"outputs":[],"source":["sc.set_figure_params(scanpy=True, dpi=150,figsize=(10,10))\n","plt.rcParams['figure.facecolor'] = 'white'\n","sc.pl.umap(adata,color=['leiden_1_0','sample'],s=0.1,ncols=1,legend_loc='on data',save='CRC_leiden_1_0.pdf')"]},{"cell_type":"code","execution_count":null,"id":"675464ed","metadata":{"id":"675464ed"},"outputs":[],"source":["sc.set_figure_params(scanpy=True, dpi=150,figsize=(6,4))\n","plt.rcParams['figure.facecolor'] = 'white'\n","sc.pl.umap(adata,color=['total_counts'],s=0.1,ncols=1,vmax='p99')"]},{"cell_type":"markdown","id":"5b8c4904","metadata":{"id":"5b8c4904"},"source":["On top of the UMAP you can visualize any variable that you might be interested (a metadata column or gene expression). We need to modify the parameter \"color\" to visualize the appropiate variable on the UMAP"]},{"cell_type":"code","execution_count":null,"id":"35fa073c","metadata":{"scrolled":true,"id":"35fa073c"},"outputs":[],"source":["keya='leiden_1_0'\n","adata.obs['X']=adata.obs['x_centroid']*4.70588\n","adata.obs['Y']=adata.obs['y_centroid']*4.70588\n","adata.obs['X_umap']=list(pd.DataFrame(adata.obsm['X_umap'])[0])\n","adata.obs['Y_umap']=list(pd.DataFrame(-adata.obsm['X_umap'])[1])\n","adata.obs['sample'].unique()\n","for rep in adata.obs['sample'].unique():\n","    adatasub=adata[adata.obs['sample']==rep]\n","    adatasub.obs[keya]=adatasub.obs[keya].astype(str)\n","    sc.pl.umap(adatasub,color=keya)\n","    import seaborn as sns\n","    map_of_clusters(adatasub,key=keya,size=4,background='white',clusters='all',save='/media/external/testserver/maria_e/CRC/CRC/clusters_tissues',figuresize=(round(adatasub.obs['X'].max()/2000),round(adatasub.obs['Y'].max()/2000)))\n"]},{"cell_type":"markdown","id":"99354890","metadata":{"id":"99354890"},"source":["# Subclustering of non-neoplastic clusters"]},{"cell_type":"code","execution_count":null,"id":"856789ce","metadata":{"id":"856789ce"},"outputs":[],"source":["interesting_clusters = ['8','18','6','10','16','13','14','4','2','17','11','7','12']"]},{"cell_type":"code","execution_count":null,"id":"d435ec11","metadata":{"id":"d435ec11"},"outputs":[],"source":["adata.obs.leiden_1_0"]},{"cell_type":"code","execution_count":null,"id":"3479308d","metadata":{"id":"3479308d"},"outputs":[],"source":["subcluster_data = adata[adata.obs['leiden_1_0'].astype(str).isin(interesting_clusters)].copy()"]},{"cell_type":"code","execution_count":null,"id":"333d348d","metadata":{"id":"333d348d"},"outputs":[],"source":["subcluster_data"]},{"cell_type":"code","execution_count":null,"id":"2324adcc","metadata":{"id":"2324adcc"},"outputs":[],"source":["sc.pp.neighbors(subcluster_data, n_neighbors=7, n_pcs=0)\n","sc.tl.umap(subcluster_data,min_dist=0.1)\n","sc.tl.leiden(subcluster_data,resolution=2.0,key_added='leiden_2_0_nei7')"]},{"cell_type":"code","execution_count":null,"id":"87c15ed1","metadata":{"id":"87c15ed1"},"outputs":[],"source":["subcluster_data.write(outpath+'/subcluster_data_after_leiden_neigh7_res2_noscale.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"061697fd","metadata":{"id":"061697fd"},"outputs":[],"source":["subcluster_data=sc.read('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_after_leiden_neigh7_res2_noscale.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"164126bc","metadata":{"id":"164126bc"},"outputs":[],"source":["sc.set_figure_params(scanpy=True, dpi=150,figsize=(10,10))\n","plt.rcParams['figure.facecolor'] = 'white'\n","sc.pl.umap(subcluster_data,color=['leiden_2_0_nei7','sample'],s=0.1,ncols=1,legend_loc='on data',save='subcluster_CRC_nonneoplastic_leiden_2_0_nei7.pdf')"]},{"cell_type":"markdown","id":"8cacb463","metadata":{"id":"8cacb463"},"source":["# Calculating frequencies"]},{"cell_type":"code","execution_count":null,"id":"9a1bc540","metadata":{"id":"9a1bc540"},"outputs":[],"source":["ctb=pd.crosstab(subcluster_data.obs['leiden_2_0_nei7'],subcluster_data.obs['sample'])"]},{"cell_type":"code","execution_count":null,"id":"3b6276ca","metadata":{"id":"3b6276ca"},"outputs":[],"source":["subcluster_data.obs"]},{"cell_type":"code","execution_count":null,"id":"d663835e","metadata":{"id":"d663835e"},"outputs":[],"source":["keya"]},{"cell_type":"code","execution_count":null,"id":"190ed84d","metadata":{"scrolled":true,"id":"190ed84d"},"outputs":[],"source":["sc.tl.rank_genes_groups(subcluster_data, groupby='leiden_2_0_nei7', method='wilcoxon')\n","sc.tl.dendrogram(subcluster_data,groupby='leiden_2_0_nei7')\n","sc.pl.rank_genes_groups_dotplot(subcluster_data, n_genes=3, swap_axes=True,save='CRC_leiden_0_nei7deg_dotplot.pdf')"]},{"cell_type":"code","execution_count":null,"id":"d4f2ee6f","metadata":{"id":"d4f2ee6f"},"outputs":[],"source":["sc.pl.rank_genes_groups(subcluster_data, n_genes=5, swap_axes=True,fontsize=20,save='CRC_leiden2_0_nei7deg.pdf')"]},{"cell_type":"code","execution_count":null,"id":"e45830cc","metadata":{"id":"e45830cc"},"outputs":[],"source":["keya='leiden_2_0_nei7'\n","subcluster_data.obs['X']=subcluster_data.obs['x_centroid']*4.70588\n","subcluster_data.obs['Y']=subcluster_data.obs['y_centroid']*4.70588\n","subcluster_data.obs['X_umap']=list(pd.DataFrame(subcluster_data.obsm['X_umap'])[0])\n","subcluster_data.obs['Y_umap']=list(pd.DataFrame(-subcluster_data.obsm['X_umap'])[1])\n","subcluster_data.obs['sample'].unique()\n","for rep in subcluster_data.obs['sample'].unique():\n","    subcluster_datasub=subcluster_data[subcluster_data.obs['sample']==rep]\n","    subcluster_datasub.obs[keya]=subcluster_datasub.obs[keya].astype(str)\n","    sc.pl.umap(subcluster_datasub,color=keya)\n","    import seaborn as sns\n","    map_of_clusters(subcluster_datasub,key=keya,size=4,background='white',clusters='all',save='/media/external/testserver/maria_e/CRC/CRC/clusters_tissues',figuresize=(round(subcluster_datasub.obs['X'].max()/2000),round(subcluster_datasub.obs['Y'].max()/2000)))"]},{"cell_type":"markdown","id":"9edc1b7e","metadata":{"id":"9edc1b7e"},"source":["# Annotating clusters"]},{"cell_type":"markdown","id":"072748a3","metadata":{"id":"072748a3"},"source":["Please input a csv with a 2 columns, one for the cluster number and one for the cell type name that you want to give to each cluster. Add the path to the csv below"]},{"cell_type":"code","execution_count":null,"id":"59b7e142","metadata":{"id":"59b7e142"},"outputs":[],"source":["class_ann=pd.read_csv(r'/media/external/testserver/maria_e/CRC/CRC/xenium_processing/annotaded_subclusters_noneoplastic.csv',sep=';')"]},{"cell_type":"code","execution_count":null,"id":"6245bb1f","metadata":{"scrolled":true,"id":"6245bb1f"},"outputs":[],"source":["class_ann"]},{"cell_type":"code","execution_count":null,"id":"c1906563","metadata":{"id":"c1906563"},"outputs":[],"source":["cls=dict(zip(class_ann['Cluster'],class_ann['Interpretation']))\n","subcluster_data.obs['cell type']=subcluster_data.obs['leiden_2_0_nei7'].astype(int).map(cls)"]},{"cell_type":"code","execution_count":null,"id":"fc4681a2","metadata":{"id":"fc4681a2"},"outputs":[],"source":["subcluster_data"]},{"cell_type":"code","execution_count":null,"id":"c0639804","metadata":{"id":"c0639804"},"outputs":[],"source":["subcluster_data.obs"]},{"cell_type":"code","execution_count":null,"id":"6d3fa934","metadata":{"scrolled":true,"id":"6d3fa934"},"outputs":[],"source":["subcluster_data.obs.leiden_2_0_nei7"]},{"cell_type":"code","execution_count":null,"id":"3c3e8cfe","metadata":{"id":"3c3e8cfe"},"outputs":[],"source":["sc.pl.umap(subcluster_data,color='cell type')"]},{"cell_type":"code","execution_count":null,"id":"98ad5d4b","metadata":{"id":"98ad5d4b"},"outputs":[],"source":["subcluster_data.write(outpath+'/subcluster_data_annotated.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"376547ae","metadata":{"id":"376547ae"},"outputs":[],"source":["import anndata\n","\n","# Load the concatenated AnnData object\n","concatenated_path = \"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_annotated_obsm.h5ad\"\n","adata_concatenated = sc.read_h5ad(concatenated_path)\n","\n","# Get unique sample tags\n","sample_tags = adata_concatenated.obs['sample'].unique()\n","\n","# Split the concatenated AnnData object into individual AnnData objects\n","individual_adatas = {}\n","for sample_tag in sample_tags:\n","    mask = (adata_concatenated.obs['sample'] == sample_tag)\n","    individual_adatas[sample_tag] = adata_concatenated[mask].copy()"]},{"cell_type":"code","execution_count":null,"id":"995479d7","metadata":{"id":"995479d7"},"outputs":[],"source":["adata_concatenated.obs['sample']"]},{"cell_type":"code","execution_count":null,"id":"c56c7ddd","metadata":{"id":"c56c7ddd"},"outputs":[],"source":["individual_adatas['0021489_Region_1'].write((\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/subclusters__noneoplastic_sample/0021489_Region_1_obsm\"+'.h5ad'))"]},{"cell_type":"code","execution_count":null,"id":"4d38abe3","metadata":{"id":"4d38abe3"},"outputs":[],"source":["individual_adatas['0021489_Region_2'].write((\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/subclusters__noneoplastic_sample/0021489_Region_2_obsm\"+'.h5ad'))"]},{"cell_type":"code","execution_count":null,"id":"8cf36005","metadata":{"id":"8cf36005"},"outputs":[],"source":["individual_adatas['0021492_Region_1'].write((\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/subclusters__noneoplastic_sample/0021492_Region_1_obsm\"+'.h5ad'))"]},{"cell_type":"code","execution_count":null,"id":"de5afe2e","metadata":{"id":"de5afe2e"},"outputs":[],"source":["individual_adatas['0021492_Region_2'].write((\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/subclusters__noneoplastic_sample/0021492_Region_2_obsm\"+'.h5ad'))"]},{"cell_type":"markdown","id":"38fc1a2d","metadata":{"id":"38fc1a2d"},"source":["# Create obsm"]},{"cell_type":"code","execution_count":null,"id":"2d460b41","metadata":{"id":"2d460b41"},"outputs":[],"source":["adata_human=sc.read(\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_annotated.h5ad\")"]},{"cell_type":"code","execution_count":null,"id":"6267ac98","metadata":{"id":"6267ac98"},"outputs":[],"source":["adata_human"]},{"cell_type":"code","execution_count":null,"id":"a764a177","metadata":{"id":"a764a177"},"outputs":[],"source":["# Load the data\n","sub_main_clusters=sc.read('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/combined_clustering.h5ad')\n","# Add spatial coordinates in obsm\n","sub_main_clusters.obsm['spatial'] = sub_main_clusters.obs[['x_centroid', 'y_centroid']].to_numpy()\n","\n","#Remove xc and yc from obs (OPTIONAL)\n","#data.obs = data.obs.drop(columns=['xc', 'yc'])\n","# Write to new h5ad\n","sub_main_clusters.write('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/combined_clustering_obsm'+'.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"b13a027f","metadata":{"id":"b13a027f"},"outputs":[],"source":["sub_main_clusters"]},{"cell_type":"code","execution_count":null,"id":"d3c655b7","metadata":{"id":"d3c655b7"},"outputs":[],"source":["adata_human=sc.read(\"/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_annotated_obsm.h5ad\")"]},{"cell_type":"markdown","id":"e2cfa896","metadata":{"id":"e2cfa896"},"source":["# Converting Xenium(morphology_mip.ome.tif) DAPI to TIFF"]},{"cell_type":"code","execution_count":null,"id":"ad5b2c99","metadata":{"id":"ad5b2c99"},"outputs":[],"source":["pip install tifffile"]},{"cell_type":"code","execution_count":null,"id":"10512ebf","metadata":{"id":"10512ebf"},"outputs":[],"source":["pip install imagecodecs"]},{"cell_type":"code","execution_count":null,"id":"f30f1221","metadata":{"id":"f30f1221"},"outputs":[],"source":["import seaborn as sns\n","import matplotlib.pyplot as plt\n","import os\n","import numpy as np\n","import pandas as pd\n","import scanpy as sc\n","import seaborn as sns\n","import gzip\n","import shutil\n","import os.path\n","from scipy.io import mmread\n","import tifffile as tf\n","from scipy.spatial import ConvexHull, convex_hull_plot_2d\n","import json\n","from PIL import Image\n","import os\n","#import PIL\n","\n","\n","path='/media/external/testserver/maria_e/CRC/CRC/Xenium/output-XETG00213__0021492__Region_2__20240125__150400' # path to the xenium folder of your input data\n","outpath='/media/external/testserver/maria_e/CRC/CRC/Xenium/DAPIS/dapi_21492_Region2' # output path. where is your file stored and how is it called? i.e media/documents/image.tif\n","def format_background(path):\n","    IM=tf.TiffFile(path+'/morphology_mip.ome.tif')\n","    position1_series = IM.series[0]\n","    position1_series.axes\n","    position1 = position1_series.asarray()\n","    return position1\n","\n","IM=format_background(path)\n","IM=IM.astype('int16')\n","im=Image.fromarray(IM)\n","#im=Image.open(path)\n","division=8 # how many times you want to\n","#im.mode='I'\n","i = im.size   # current size (height,width)\n","i = i[0]//division, i[1]//division  # new size\n","newimage = im.resize(i)\n","\n","newimage.save(outpath+'.tif')"]},{"cell_type":"markdown","id":"7ed1781c","metadata":{"id":"7ed1781c"},"source":["# Join subclustering+annotations to the main clustering anndata"]},{"cell_type":"code","execution_count":null,"id":"3d01187a","metadata":{"id":"3d01187a"},"outputs":[],"source":["import seaborn as sns\n","import matplotlib.pyplot as plt\n","import os\n","import numpy as np\n","import pandas as pd\n","import scanpy as sc\n","import seaborn as sns\n","import gzip\n","import shutil\n","import os.path\n","from scipy.io import mmread\n","import tifffile as tf\n","from scipy.spatial import ConvexHull, convex_hull_plot_2d\n","import json\n","from PIL import Image\n","import os\n","import anndata\n","# Load your initial AnnData object\n","initial_anndata = anndata.read_h5ad('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/adata_after_leiden_noscale_obsm.h5ad')\n","\n","# Load the subclusters and annotations\n","subclusters_anndata = anndata.read_h5ad('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_annotated_obsm.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"b76aad1f","metadata":{"id":"b76aad1f"},"outputs":[],"source":["initial_anndata"]},{"cell_type":"code","execution_count":null,"id":"2cecbd8c","metadata":{"scrolled":true,"id":"2cecbd8c"},"outputs":[],"source":["subclusters_anndata"]},{"cell_type":"code","execution_count":null,"id":"fb28bb27","metadata":{"id":"fb28bb27"},"outputs":[],"source":["initial_anndata.obs['unique_id']=initial_anndata.obs['cell_id'].astype(str)+'_'+initial_anndata.obs['sample'].astype(str)\n","subclusters_anndata.obs['unique_id']=subclusters_anndata.obs['cell_id'].astype(str)+'_'+subclusters_anndata.obs['sample'].astype(str)"]},{"cell_type":"code","execution_count":null,"id":"a96f9afc","metadata":{"id":"a96f9afc"},"outputs":[],"source":["dictionary_id_to_celltype=dict(zip(subclusters_anndata.obs['unique_id'],subclusters_anndata.obs['cell type'])) #zip is for pairing one column with another"]},{"cell_type":"code","execution_count":null,"id":"5c99f5b7","metadata":{"id":"5c99f5b7"},"outputs":[],"source":["initial_anndata.obs['combined_clusters']=initial_anndata.obs['unique_id'].map(dictionary_id_to_celltype)"]},{"cell_type":"code","execution_count":null,"id":"b2003740","metadata":{"id":"b2003740"},"outputs":[],"source":["initial_anndata.obs['combined_clusters']=initial_anndata.obs['combined_clusters'].astype(str)"]},{"cell_type":"code","execution_count":null,"id":"8b54e641","metadata":{"id":"8b54e641"},"outputs":[],"source":["initial_anndata.obs.loc[initial_anndata.obs['combined_clusters']=='nan','combined_clusters']='Neoplastic'#loc for selecting rows and columns,first part is row, and second ('combined_clusters') is columns"]},{"cell_type":"code","execution_count":null,"id":"67039ddd","metadata":{"id":"67039ddd"},"outputs":[],"source":["sc.pl.umap(initial_anndata,color='combined_clusters')"]},{"cell_type":"code","execution_count":null,"id":"e7d61c12","metadata":{"id":"e7d61c12"},"outputs":[],"source":["initial_anndata.write_h5ad('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/combined_clustering.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"ad3a3d30","metadata":{"id":"ad3a3d30"},"outputs":[],"source":["import seaborn as sns\n","import matplotlib.pyplot as plt\n","import os\n","import numpy as np\n","import pandas as pd\n","import scanpy as sc\n","import seaborn as sns\n","import gzip\n","import shutil\n","import os.path\n","from scipy.io import mmread\n","import tifffile as tf\n","from scipy.spatial import ConvexHull, convex_hull_plot_2d\n","import json\n","from PIL import Image\n","import os\n","import anndata"]},{"cell_type":"code","execution_count":null,"id":"ec580319","metadata":{"id":"ec580319"},"outputs":[],"source":["patata = anndata.read_h5ad('/media/external/testserver/maria_e/CRC/adata_with_factors.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"a62b89e7","metadata":{"id":"a62b89e7"},"outputs":[],"source":["patata"]},{"cell_type":"code","execution_count":null,"id":"992328c0","metadata":{"id":"992328c0"},"outputs":[],"source":["sc.tl.rank_genes_groups(patata, groupby='combined_clusters', method='wilcoxon')\n","sc.tl.dendrogram(patata,groupby='combined_clusters')\n","sc.pl.rank_genes_groups_dotplot(patata, n_genes=3, swap_axes=True,save='CRC_240724_dotplot.pdf')"]},{"cell_type":"code","execution_count":null,"id":"d12df428","metadata":{"id":"d12df428"},"outputs":[],"source":["#to get the curremt working directory where the saved files are being saved when you only type'save' and nothing else. without path\n","import os\n","print(os.getcwd())\n"]},{"cell_type":"code","execution_count":null,"id":"e2b77c2e","metadata":{"id":"e2b77c2e"},"outputs":[],"source":["sc.pl.rank_genes_groups(patata, n_genes=5, swap_axes=True,fontsize=20,save='CRC_final_ranking.pdf')"]},{"cell_type":"markdown","id":"dfa214c4","metadata":{"id":"dfa214c4"},"source":["# Reading the object again if needed"]},{"cell_type":"markdown","id":"4c31d990","metadata":{"id":"4c31d990"},"source":["# Cell type consistency between replicates"]},{"cell_type":"code","execution_count":null,"id":"1cb19c7d","metadata":{"id":"1cb19c7d"},"outputs":[],"source":["keya='cell type'"]},{"cell_type":"code","execution_count":null,"id":"8c61699b","metadata":{"id":"8c61699b"},"outputs":[],"source":["crs=pd.crosstab(adata.obs[keya],adata.obs['sample'])\n","crs_perc=crs.div(crs.sum(axis=0),axis=1)*100\n","#crs_perc['celltype_annotation']=crs_perc.index\n","crs_transpose=crs_perc.transpose()\n","crs_transpose['sample']=crs_transpose.index\n","import matplotlib\n","matplotlib.rcParams['pdf.fonttype'] = 42\n","matplotlib.rcParams['ps.fonttype'] = 42\n","crs_transpose.set_index('sample').plot(kind='bar', stacked=True, color=adata.uns[keya+'_colors'])\n","#plt.savefig('../../figures/celltype_identification_mscoronal/stacked_barplot_celltypes.pdf')"]},{"cell_type":"code","execution_count":null,"id":"6fd764a2","metadata":{"scrolled":true,"id":"6fd764a2"},"outputs":[],"source":["keya='cell type'\n","adata.obs['X']=adata.obs['x_centroid']*4.70588\n","adata.obs['Y']=adata.obs['y_centroid']*4.70588\n","adata.obs['X_umap']=list(pd.DataFrame(adata.obsm['X_umap'])[0])\n","adata.obs['Y_umap']=list(pd.DataFrame(-adata.obsm['X_umap'])[1])\n","adata.obs['sample'].unique()\n","for rep in adata.obs['sample'].unique():\n","    adatasub=adata[adata.obs['sample']==rep]\n","    adatasub.obs[keya]=adatasub.obs[keya].astype(str)\n","    sc.pl.umap(adatasub,color=keya)\n","    import seaborn as sns\n","    map_of_clusters(adatasub,key=keya,size=4,background='white',clusters='all',figuresize=(round(adatasub.obs['X'].max()/2000),round(adatasub.obs['Y'].max()/2000)))"]},{"cell_type":"markdown","id":"a1782416","metadata":{"id":"a1782416"},"source":["# EXPLORATION OF SOME CLUSTERS"]},{"cell_type":"code","execution_count":null,"id":"e77b0799","metadata":{"id":"e77b0799"},"outputs":[],"source":["import scanpy as sc\n","subcluster_data=sc.read('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/subcluster_data_annotated_obsm.h5ad')\n"]},{"cell_type":"code","execution_count":null,"id":"b6efa322","metadata":{"id":"b6efa322"},"outputs":[],"source":["sc.pl.umap(subcluster_data,color=['TIMP3','IGFBP7','THBS1','VCAN','DPYSL3','MAF'],vmax='p99',ncols=2)#display as percintile 99"]},{"cell_type":"code","execution_count":null,"id":"34adf92e","metadata":{"id":"34adf92e"},"outputs":[],"source":["genes_to_plot=['MAF']#,'IGFBP7','THBS1','VCAN','DPYSL3','MAF']\n","for gene in genes_to_plot:\n","    for s in subcluster_data.obs['sample'].unique():\n","        sub=subcluster_data[subcluster_data.obs['sample']==s]\n","        sc.pl.spatial(sub,color=gene,vmax='p99',spot_size=40)#display as percintile 99"]},{"cell_type":"code","execution_count":null,"id":"80206c72","metadata":{"id":"80206c72"},"outputs":[],"source":["list(subcluster_data.obs['cell type'].unique())"]},{"cell_type":"code","execution_count":null,"id":"09ed7942","metadata":{"id":"09ed7942"},"outputs":[],"source":["selected_clusters=['endothelial/fibroblast_4',\n"," 'endothelial/fibroblast_2',\n"," 'endothelial_3',\n"," 'endothelial/fibroblast_1','endothelial/fibroblast_3','T cell?_3']#,'IGFBP7','THBS1','VCAN','DPYSL3','MAF']\n","adata_fibro=subcluster_data[subcluster_data.obs['cell type'].isin(selected_clusters)]"]},{"cell_type":"code","execution_count":null,"id":"466ade97","metadata":{"id":"466ade97"},"outputs":[],"source":["#Differential expression of genes between the selcted clusters\n","sc.tl.rank_genes_groups(adata_fibro, groupby='cell type', method='wilcoxon')\n","sc.tl.dendrogram(adata_fibro,groupby='cell type')\n","sc.pl.rank_genes_groups_dotplot(adata_fibro, n_genes=5, swap_axes=True,save='CRC_final_dotplot_only_fibro_subtypes.pdf')"]},{"cell_type":"code","execution_count":null,"id":"a0a65715","metadata":{"id":"a0a65715"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"id":"3dbb3322","metadata":{"id":"3dbb3322"},"outputs":[],"source":["crsnorm=crs.div(crs.sum(axis=0),axis=1)"]},{"cell_type":"code","execution_count":null,"id":"1050b29d","metadata":{"id":"1050b29d"},"outputs":[],"source":["crsnorm=crsnorm.div(crsnorm.sum(axis=1),axis=0)\n","crsnorm=crsnorm.loc[:,['0021489_Region_1','0021492_Region_1', '0021489_Region_2', '0021492_Region_2']]"]},{"cell_type":"code","execution_count":null,"id":"eed738e6","metadata":{"id":"eed738e6"},"outputs":[],"source":["plt.figure(figsize=(15,15))\n","crsnorm.plot(kind='barh', stacked=True,color=['black','purple','yellow','orange'],width=1)\n","plt.grid('off')"]},{"cell_type":"markdown","id":"f06ae559","metadata":{"id":"f06ae559"},"source":["# Squidpy analysis"]},{"cell_type":"code","execution_count":null,"id":"14ea197f","metadata":{"id":"14ea197f"},"outputs":[],"source":["adata.obsm['spatial']=np.array(adata.obs.loc[:,['X','Y']])"]},{"cell_type":"code","execution_count":null,"id":"2b6fe905","metadata":{"id":"2b6fe905"},"outputs":[],"source":["anndata_list = []\n","for sample in adata.obs['sample'].unique():\n","    print(sample)\n","    adata_copy_int = adata[adata.obs['sample']==sample]\n","    sq.gr.spatial_neighbors(adata_copy_int,radius=124,coord_type = 'generic')\n","    anndata_list.append(adata_copy_int)\n","ad_sp_concat = sc.concat(anndata_list, pairwise=True)\n","ad_sp_concat.uns=adata.uns"]},{"cell_type":"code","execution_count":null,"id":"c047a6c8","metadata":{"id":"c047a6c8"},"outputs":[],"source":["ad_sp_concat.obs['cell type']=ad_sp_concat.obs['cell type'].astype('category')"]},{"cell_type":"code","execution_count":null,"id":"8eef0710","metadata":{"id":"8eef0710"},"outputs":[],"source":["level='cell type'\n","nenrich=sq.gr.nhood_enrichment(ad_sp_concat, cluster_key=level,copy = True)\n","sq.gr.nhood_enrichment(ad_sp_concat, cluster_key=level)\n","#ad_sp_concat.uns[level+'_nhood_enrichment']['zscore']=np.array(pd.DataFrame(ad_sp_concat.uns[level+'_nhood_enrichment']['zscore']).fillna(0))"]},{"cell_type":"code","execution_count":null,"id":"68b1f7cd","metadata":{"id":"68b1f7cd"},"outputs":[],"source":["import matplotlib\n","sc.set_figure_params(scanpy=True, dpi=200, dpi_save=200, frameon=False, vector_friendly=True, fontsize=7, figsize=(8,8), color_map=None, format='pdf', facecolor=None, transparent=False, ipython_format='png2x')\n","matplotlib.rcParams['pdf.fonttype'] = 42\n","matplotlib.rcParams['ps.fonttype'] = 42\n","sq.pl.nhood_enrichment(ad_sp_concat, cluster_key=level, method=\"ward\",mode='zscore',vmax=120,vmin=-120,cmap='coolwarm',show=False)\n","#plt.savefig(general_path+'/ISS_global_analysis/global_nhood_enrichment_'+level+'.pdf')"]},{"cell_type":"markdown","id":"521c5979","metadata":{"id":"521c5979"},"source":["# CORRECT TUMORS"]},{"cell_type":"code","execution_count":null,"id":"53f629aa","metadata":{"id":"53f629aa"},"outputs":[],"source":["adata.obs['sample'].unique()"]},{"cell_type":"code","execution_count":null,"id":"1e075441","metadata":{"id":"1e075441"},"outputs":[],"source":["adata_neo=adata[adata.obs['cell type']=='Neoplastic'].copy()"]},{"cell_type":"code","execution_count":null,"id":"1abc94bd","metadata":{"id":"1abc94bd"},"outputs":[],"source":["adata89_r1=adata_neo[adata_neo.obs['sample']=='0021489_Region_2']"]},{"cell_type":"code","execution_count":null,"id":"ddd22e44","metadata":{"id":"ddd22e44"},"outputs":[],"source":["sc.pp.neighbors(adata89_r1, n_pcs =0)\n","sc.tl.umap(adata89_r1)\n","sc.tl.leiden(adata89_r1,key_added='new_leiden')"]},{"cell_type":"code","execution_count":null,"id":"d2ff60b1","metadata":{"id":"d2ff60b1"},"outputs":[],"source":["sc.pl.umap(adata89_r1,color='new_leiden',s=10)"]},{"cell_type":"code","execution_count":null,"id":"e5419e7e","metadata":{"id":"e5419e7e"},"outputs":[],"source":["sc.pl.umap(adata89_r1,color='leiden_2_0',s=10)"]},{"cell_type":"code","execution_count":null,"id":"c6ce5a22","metadata":{"id":"c6ce5a22"},"outputs":[],"source":["keya='new_leiden'\n","sc.tl.rank_genes_groups(adata89_r1, groupby=keya, method='wilcoxon')\n","sc.tl.dendrogram(adata89_r1,groupby=keya)\n","sc.pl.rank_genes_groups_dotplot(adata89_r1, n_genes=6, swap_axes=True)#,save='deg.pdf')"]},{"cell_type":"code","execution_count":null,"id":"62ee4eb9","metadata":{"id":"62ee4eb9"},"outputs":[],"source":["keya='new_leiden'\n","adata89_r1.obs['X_umap']=list(pd.DataFrame(adata89_r1.obsm['X_umap'])[0])\n","adata89_r1.obs['Y_umap']=list(pd.DataFrame(-adata89_r1.obsm['X_umap'])[1])\n","for rep in adata89_r1.obs['sample'].unique():\n","    adatasub=adata89_r1[adata89_r1.obs['sample']==rep]\n","    adatasub.obs[keya]=adatasub.obs[keya].astype(str)\n","    sc.pl.umap(adatasub,color=keya)\n","    import seaborn as sns\n","    map_of_clusters(adatasub,key=keya,size=4,background='white',clusters='all',figuresize=(round(adatasub.obs['X'].max()/2000),round(adatasub.obs['Y'].max()/2000)))"]},{"cell_type":"markdown","id":"3d56fd5d","metadata":{"id":"3d56fd5d"},"source":["# Correct tumors"]},{"cell_type":"code","execution_count":null,"id":"a4518b66","metadata":{"id":"a4518b66"},"outputs":[],"source":["adata_neo=adata[adata.obs['cell type']=='Neoplastic'].copy()"]},{"cell_type":"code","execution_count":null,"id":"962fa11e","metadata":{"id":"962fa11e"},"outputs":[],"source":["adata_neo"]},{"cell_type":"code","execution_count":null,"id":"83bae2ba","metadata":{"id":"83bae2ba"},"outputs":[],"source":["sc.pp.combat(adata_neo,key='sample')"]},{"cell_type":"code","execution_count":null,"id":"9603ac57","metadata":{"id":"9603ac57"},"outputs":[],"source":["adata_neo"]},{"cell_type":"code","execution_count":null,"id":"9c1c7062","metadata":{"id":"9c1c7062"},"outputs":[],"source":["sc.pp.neighbors(adata_neo, n_pcs =0)\n","sc.tl.umap(adata_neo)\n","sc.tl.leiden(adata_neo,key_added='new_leiden')"]},{"cell_type":"code","execution_count":null,"id":"5be960f7","metadata":{"id":"5be960f7"},"outputs":[],"source":["sc.pl.umap(adata_neo,color='sample',s=2)"]},{"cell_type":"code","execution_count":null,"id":"3a76b29f","metadata":{"id":"3a76b29f"},"outputs":[],"source":["sc.pl.umap(adata_neo,color='leiden_2_0',s=10)"]},{"cell_type":"code","execution_count":null,"id":"3c225c41","metadata":{"id":"3c225c41"},"outputs":[],"source":["sc.pl.umap(adata_neo,color='new_leiden',s=10)"]},{"cell_type":"code","execution_count":null,"id":"6df64463","metadata":{"id":"6df64463"},"outputs":[],"source":["keya='new_leiden'\n","adata_neo.obs['X_umap']=list(pd.DataFrame(adata_neo.obsm['X_umap'])[0])\n","adata_neo.obs['Y_umap']=list(pd.DataFrame(-adata_neo.obsm['X_umap'])[1])\n","for rep in adata_neo.obs['sample'].unique():\n","    adatasub=adata_neo[adata_neo.obs['sample']==rep]\n","    adatasub.obs[keya]=adatasub.obs[keya].astype(str)\n","    sc.pl.umap(adatasub,color=keya)\n","    import seaborn as sns\n","    map_of_clusters(adatasub,key=keya,size=4,background='white',clusters='all',figuresize=(round(adatasub.obs['X'].max()/2000),round(adatasub.obs['Y'].max()/2000)))"]},{"cell_type":"markdown","id":"ec0d5c9a","metadata":{"id":"ec0d5c9a"},"source":["# What's inside of the Anndata"]},{"cell_type":"code","execution_count":null,"id":"cdace072","metadata":{"id":"cdace072"},"outputs":[],"source":["adata=sc.read('/media/external/testserver/maria_e/CRC/CRC/xenium_processing/combined_adata/combined_clustering_obsm.h5ad')"]},{"cell_type":"code","execution_count":null,"id":"7c338a3e","metadata":{"id":"7c338a3e"},"outputs":[],"source":["adata"]},{"cell_type":"code","execution_count":null,"id":"8da24a88","metadata":{"id":"8da24a88"},"outputs":[],"source":["import numpy as np\n","import pandas as pd\n","import scanpy as sc\n","\n","# Assuming your Anndata object is named 'adata'\n","# If you haven't already installed scanpy, you can do so via pip:\n","# !pip install scanpy\n","\n","# Convert the data matrix to a DataFrame for easier manipulation\n","gene_counts_df = pd.DataFrame(adata.X, columns=adata.var_names)\n","\n","# Now you can display the number of counts per gene\n","gene_counts = gene_counts_df.sum(axis=0)\n","\n","# Display the counts per gene\n","print(gene_counts)\n"]},{"cell_type":"code","execution_count":null,"id":"dfcd33d2","metadata":{"id":"dfcd33d2"},"outputs":[],"source":["import scanpy as sc"]},{"cell_type":"code","execution_count":null,"id":"2e554898","metadata":{"id":"2e554898"},"outputs":[],"source":["import pandas as pd\n","\n","# Assuming your Anndata object is named 'adata'\n","\n","# Convert the data matrix to a DataFrame for easier manipulation\n","transcript_counts_df = pd.DataFrame(adata.X, columns=adata.var_names, index=adata.obs['cell_id'])\n","\n","# Now you can display the number of transcripts per gene per cell\n","print(transcript_counts_df)\n"]},{"cell_type":"code","execution_count":null,"id":"b01db1ee","metadata":{"id":"b01db1ee"},"outputs":[],"source":["transcript_counts_df = pd.DataFrame(adata.X)"]},{"cell_type":"code","execution_count":null,"id":"9971957a","metadata":{"id":"9971957a"},"outputs":[],"source":["transcript_counts_df"]},{"cell_type":"code","execution_count":null,"id":"9b5c40c8","metadata":{"id":"9b5c40c8"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.4"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}